packages/optimizer/lib/transformers/PreloadHeroImage.js

```diff
@@ -0,0 +1,196 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const {createElement, nextNode, insertAfter, firstChildByTag} = require('../NodeUtils');
+const {findMetaViewport} = require('../HtmlDomHelper');
+const parseSrcSet = require('../parseSrcSet');
+
+// Images smaller than 150px are considered tiny
+const TINY_IMG_THRESHOLD = 150;
+
+/**
+ * PreloadHeroImage - This transformer identifies a hero image or
+ * important images on the document and attaches <link rel="preload"> element.
+ *
+ * This transformer supports the following option:
+ *
+ * * `preloadHeroImage`: [true|false] - enables or disables hero image preloading. The default is `true`.
+ */
+class PreloadHeroImage {
+  constructor(config) {
+    this.log = config.log;
+    this.enabled = config.preloadHeroImage !== false && config.experimentPreloadHeroImage;
+    this.log.info('enabled', this.enabled);
+  }
+  async transform(root, params) {
+    if (!this.enabled || params.preloadHeroImage === false) {
+      return;
+    }
+    const html = firstChildByTag(root, 'html');
+    const head = firstChildByTag(html, 'head');
+    const body = firstChildByTag(html, 'body');
+    if (!body || !head) return;
+
+    const heroImage = this.findHeroImage(body);
+
+    if (!heroImage) {
+      return;
+    }
+    let referenceNode = findMetaViewport(head);
+
+    const preload = createElement('link', {
+      'rel': 'preload',
+      'href': heroImage.src,
+      'as': 'image',
+      'data-hero': '',
+    });
+    if (heroImage.srcset) {
+      try {
+        parseSrcSet(heroImage.srcset);
+        preload.attribs.srcset = heroImage.srcset;
+      } catch (err) {
+        this.log.warn(
+          `Invalid srcset '${heroImage.srcset}' for amp-img src '${heroImage.src}'`,
+          err.message
+        );
+      }
+    }
+    this.log.debug('Preloading hero image: ', heroImage.src);
+    insertAfter(head, preload, referenceNode);
+  }
+
+  findHeroImage(root) {
+    if (!root.tagName) {
+      return null;
+    }
+    // Ignore images inside templates
+    if (root.tagName === 'template') {
+      return null;
+    }
+
+    const layout = root.attribs ? root.attribs.layout : '';
+    if (layout === 'nodisplay') return null;
+
+    if (root.tagName === 'amp-img') {
+      return this.isCandidateImageForPreloading(root);
+    }
+    if (root.tagName === 'amp-video') {
+      return this.isCandidateVideoPosterImage(root);
+    }
+    if (root.tagName === 'amp-iframe' || root.tagName === 'amp-video-iframe') {
+      return this.isCandidateIframePlaceholderImage(root);
+    }
```

Comment:
would it make sense to check for background image on large elements?

Reply:
What would be a good way to find large (non-amp elements?

Reply:
[heres](https://gist.github.com/patrickkettner/ad496a6acbaa3b28dbe1fb88aeb25a9e) what we discussed during the standup

Reply:
Thanks! How could this be integrated into the transformer as it doesn't run in the browser?

Reply:
you'd need to do something like shell out to a headless browser, since domutils doesn't seem to touch any stylesheets (and therefore can't get rect sizes or background images). 

so nevermind 🙃